stdlib/Markdown/src/render/terminal/formatting.jl

```diff
@@ -1,68 +1,65 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-# Wrapping
+const AnnotIO = Union{AnnotatedIOBuffer, IOContext{AnnotatedIOBuffer}}
 
-function ansi_length(s)
-    replace(s, r"\e\[[0-9]+m" => "") |> textwidth
+function annotprint(f::Function, args...)
+    buf = AnnotatedIOBuffer()
+    f(buf, args...)
+    read(buf, AnnotatedString)
 end
 
-words(s) = split(s, " ")
-lines(s) = split(s, "\n")
+"""
+    with_output_annotations(f::Function, io::AnnotIO, annots::Pair{Symbol, <:Any}...)
 
-function wrapped_line(io::IO, s::AbstractString, width, i)
-    ws = words(s)
-    lines = String[]
-    for word in ws
-        word_length = ansi_length(word)
-        word_length == 0 && continue
-        if isempty(lines) || i + word_length + 1 > width
-            i = word_length
-            if length(lines) > 0
-                last_line = lines[end]
-                maybe_underline = findlast(Base.text_colors[:underline], last_line)
-                if !isnothing(maybe_underline)
-                    # disable underline style at end of line if not already disabled.
-                    maybe_disable_underline = max(
-                        last(something(findlast(Base.disable_text_style[:underline], last_line), -1)),
-                        last(something(findlast(Base.text_colors[:normal], last_line), -1)),
-                    )
+Call `f(io)`, and apply `annots` to the output created by doing so.
+"""
+function with_output_annotations(f::Function, io::AnnotIO, annots::Pair{Symbol, <:Any}...)
+    @nospecialize f annots
+    aio = if io isa AnnotatedIOBuffer io else io.io end
+    start = position(aio) + 1
+    f(io)
+    stop = position(aio)
+    for annot in annots
+        push!(aio.annotations, (start:stop, annot))
```

Comment:
Does it matter if these are strictly appended (e.g. `aio.annotations` is unsorted afterwards)?

Reply:
Sorted insertion is currently not *required* at any particular, but treated as nice to have, and so I've make all of the operations on annotations in `Base` and `StyledStrings` preserve sorted ordering. It would be good to do so here.